Fix bc-wp-init.sh to work on Windows: find hosts file in Windows and …

barikamada opened the pull request:
…Lunux OS, generate random number for forwarded_port option

adionditsak commented on scripts/bc-wp-init.sh:
```diff
@@ -41,19 +41,28 @@ nextip(){
 
-# 
-if [[ -n `cat /etc/hosts | grep $domain` ]]
+# WIndows and Linux have diff paths to hosts file
```
What about Mac?

barikamada replied:
Add support for MacOs here. Thank you!

adionditsak replied:
Please fix typo capital `I` in `WIndows` and please do not use abbreviations like `diff` in comments. The whole sentence should be changed to UNIX instead of Linux or include Mac. If you want a comment then it should make sense

barikamada replied:
Remove comments, commit b6bb5e54

barikamada commented:
Script bc-wp-init.sh uses data from /etc/hosts file to choose proper ip adress for new virtualbox instance. But hosts file in Windows is located in C:\Windows\System32\drivers\etc\ directory. I added code to choose right path to hosts file.
Also script generated random numper for forward_port option using python oneliner. But often python is not setup on WIndows shell (MinGW64), and forward_port gets no value. I replaced python code with `shuf` tool, it's working fine on Windows.

adionditsak commented on scripts/bc-wp-init.sh:
```diff
@@ -41,19 +41,30 @@ nextip(){
-if [[ -n `cat /etc/hosts | grep $domain` ]]
+# WIndows and Linux have diff paths to hosts file
+case "$OSTYPE" in 
+    "linux-gnu"|"darwin")
```
Great, but please make more generic, eg. `*linux*`, `*darwin*`, `*msys*`. This would for example not work in my case:
`➜  ~ echo $OSTYPE
darwin18.0`

adionditsak replied:
`*win*` might also make sense for windows

barikamada replied:
Fixed in commit 9d68d7a

adionditsak commented on scripts/bc-wp-init.sh:
```diff
@@ -41,19 +41,32 @@ nextip(){
 sed -i -e "/^ip:/s/.*/ip: $ip/" site.yml
 
-forwarded_port=`python -c "import random; print random.randint(8000,8999)"`
+forwarded_port=`shuf -i 8000-8999 -n 1`
```
Are `shuf` supported in all operating systems? I reckon it only works for Linux

adionditsak replied:
Also, there is a chance that the port, when randomly chosen, will be the same as an existing port. Right?

barikamada replied:
> Are `shuf` supported in all operating systems? I reckon it only works for Linux

shuf works on systems which I had a possibility to test - Ubuntu and Windows. shuf is part of coreutils, odds to have coreutils on MacOs very high. So I'm sure we need to use this.
 

> Also, there is a chance that the port, when randomly chosen, will be the same as an existing port. Right?

It can be, but firstly the probability is low, then checking the fact that another virtualbox instance took the same value for forwarded_port will complicate script (and even won't solve this problem completely). So I don't have another solution here.

adionditsak replied:
Firstly, `shuf` is not supported in macos, and Mac users, which is the main part of our developers, have to install `brew` and then `coreutils` through `brew` to use `shuf`. We can not ask developers cross-office to do this, as it will complicate the process and we want the opposite. Please check this before deciding on a util. Secondly, why will it not solve the issue to check if the port is in use already? Use `netstat`, `ss` or even `vagrant port [id]` for this - I can reckon it can be resolved in a simple way

barikamada commented:
Commit 364d895: 
vagrant has an option to check if configured forwarded_port is taken on underlying host  and if so vagrant will choose another available port (https://www.vagrantup.com/docs/ networking/forwarded_ports.html). So I changed Vagrant file that way:

`config.vm.network "forwarded_port", guest: 80, host: _conf['forwarded_port'], auto_correct: true
config.vm.usable_port_range = 8181..8999`

Mentioned option is `auto_correct: true`. 
With such Vagrant file first vagrant box will have forwarding 80 (guest) => 8181 (host). Then next vagrant box will check whether 8181 is taken, will discover that this is true and will try to use next ports available from range 8181..8999 (because of option config.vm.usable_port_range).
I removed code from scripts/bc-wp-init.sh which generate value for forwarded_port - it's not needed anymore.

adionditsak commented:
@barikamada good stuff!
